cli/command/cli.go

```diff
@@ -62,6 +63,9 @@ type Cli interface {
 	DockerEndpoint() docker.Endpoint
 }
 
+// DockerCliOperator applies a modification on a DockerCli.
+type DockerCliOperator func(cli *DockerCli)
```

Comment:
perhaps `s/DockerCliOperator/DockerCliOption/` ?

note that strictly, we should've used CLI (all caps), but that ship has likely sailed 😓 


One thing I'm thinking; should we change the signature to return an `error`? This would allow us to capture possible errors in the configuration (which could be if these functional options take (e.g.) configuration from something else, and there's an issue with the configuration, or (e.g.) if there's conflicting options set


@vdemeester @ijc wdyt?

Reply:
I do agree on the "return `error`" :angel: :wink:. I do like `Operator` somehow :upside_down_face:

Reply:
👍 for the renaming.
About the error, I thought about that too, I just wondered it may modify too much code for the caller, as suddenly it has to return and propagate an error, if it didn't already. But ok I'll add that 👍

Reply:
> I think that this func typedef should certainly return an error, but I don't think it necessarily has to follow that NewDockerCli needs to propagate it for the user to handle

I've seen two patterns; one that never errors, and one that can error. I think having an error is more flexible (errors can be ignored if we don't need them, or a functional option can just return `nil`)

> but make a more dynamic decision about whether to use (and this Initialize) it further down the line.

Right, so I was thinking about that a bit; there's some options that may depend on runtime options (e.g. if a specific flag or environment variable is set) and those have to be evaluated when a command is triggered.

Functions are first-class citizens in Go; should we have something to pass functions that can be executed at runtime (e.g. for the context/config store to be evaluated when the command is run, instead of having to manually call `Initialize()` before running a command?

Reply:
Having the process exit on NewDockerCli() seem a bit dangerous to me, for 3rd party consumers. I'd vote for NewDockerCli() returning an error.

Reply:
For Post-Initialize functions, we can still use Apply()